### DISCUSSION TITLE
Execute xonsh code to modify class attributes

### DISCUSSION THREAD

**Discussion by nitanmarcel** [xonsh/xonsh#5687] [Q&A] [answered]
I was wondering if there's a way to use it as an rc (Run Commands) file, to allow configuration?

What I think I will need for this is a way to to pass a specific class that I use for configuration when loading the file, then execute it and the code would modify that class parameters.

Is something like this possible somehow?

**Comment by anki-code** [comment]
You mean [xonshrc](https://xon.sh/xonshrc.html)? It's not clear by your description of what you want to implement.

**Reply by nitanmarcel** [reply]
> You mean [xonshrc](https://xon.sh/xonshrc.html)? It's not clear by your description of what you want to implement.

For a custom application, something that I can execute at launch and it will edit the attributes of a specific class.

For example:

```python
class MyApp:
    def __init__(self):
        self.x = "Something"
    
    def init(self, rc_path):
         with open(rc_path, "r") as rc:
            xonsh.execute_code_with_some_global(globals={"app": self}, code=rc.read())
 ```

Som the xonsh code will be able to use that global to change the x value: app.x = "something else"

**Reply by anki-code** [reply]
What about using [`execx`](https://xon.sh/tutorial.html#compile-evaluate-execute)?
```xsh
class A:
    a=1
    def __init__(self):
        print(self.a)
        execx('app.a=$(echo 2)', glbs={"app":self})
        print(self.a)

a = A()
# 1
# 2
```

**Reply by nitanmarcel** [reply]
That should work thanks

**Reply by nitanmarcel** [reply]
I just tested it. Full code is:

```python
from xonsh.execer import Execer
from xonsh.built_ins import get_default_builtins
builtinsx = get_default_builtins(Execer())
builtinsx.execx('app.a=$(echo 2)', glbs={"app":self})
```

**Reply by nitanmarcel** [reply]
> What about using [`execx`](https://xon.sh/tutorial.html#compile-evaluate-execute)?
> 
> ```python
> class A:
>     a=1
>     def __init__(self):
>         print(self.a)
>         execx('app.a=$(echo 2)', glbs={"app":self})
>         print(self.a)
> 
> a = A()
> # 1
> # 2
> ```

any idea why when I do `app.a=$(echo 2)` my python program just stops?

`fish: Job 30, 'python3 src/sno_py/ test.py' has stopped`

```
There are still jobs active:

   PID  Command
343306  python3 src/sno_py/ test.py
342689  python3 src/sno_py/ test.py
342609  python3 src/sno_py/ test.py
342224  python3 src/sno_py/ test.py
335966  python3 src/sno_py/ test.py
334084  python3 src/sno_py/ test.py
333645  python3 src/sno_py/ test.py
333612  python3 src/sno_py/ test.py
332807  python3 src/sno_py/ test.py
332528  python3 src/sno_py/ test.py
331002  python3 src/sno_py/ test.py
330915  python3 src/sno_py/ test.py
330356  python3 src/sno_py/ test.py
330017  python3 src/sno_py/ test.py
328075  python3 src/sno_py/ test.py
326475  python3 src/sno_py/ test.py
326084  python3 src/sno_py/ test.py
325404  python3 src/sno_py/ test.py
325075  python3 src/sno_py/ test.py
324286  python3 src/sno_py/ test.py
323911  python3 src/sno_py/ test.py
323426  python3 src/sno_py/ test.py
323055  python3 src/sno_py/ test.py
322387  python3 src/sno_py/ test.py
321370  python3 src/sno_py/ test.py
321278  python3 src/sno_py/ test.py
320920  python3 src/sno_py/ test.py
320435  python3 src/sno_py/ test.py
320062  python3 src/sno_py/ test.py
286342  python3 src/sno_py/ test.py
286315  python3 src/sno_py/ test.py

A second attempt to exit will terminate them.
Use 'disown PID' to remove jobs from the list without terminating them.
```

**Reply by nitanmarcel** [reply]
No idea what happened, I fixed it by running it async with asyncer

**Reply by anki-code** [reply]
I think it's something about SIGTTIN, SIGTTOU (https://github.com/anki-code/xonsh-developer-toolkit/?tab=readme-ov-file#docs-1).

**Comment by nitanmarcel** [comment, marked as answer]
Posting this if someone wants to do the same thing. 

Looks like the full code I posted before was not enough. It worked fine when using python or bash commands, but as soon as you combine them (`foo=$(echo bar)`) the program will throw exceptions. That's because the whole xonsh session needs to be loaded.

The full code would be something like this

```python
import xonsh.built_ins
import xonsh.execer
import xonsh.imphooks
import xonsh.environ

execer = xonsh.execer.Execer()
xonsh.built_ins.XSH.load(execer=execer, inherit_env=True)
xonsh.imphooks.install_import_hooks(execer=execer) 
xonsh.built_ins.XSH.builtins.execx("foo=$(echo bar)", glbs={"editor": self})
```

**Comment by nitanmarcel** [comment]
If you don't mind. How it's the output of ls, cat, etc printed to the shell?

I'm redirecting stdout and stderr and I still the output in the terminal instead of the the app

https://github.com/nitanmarcel/Sno-py/blob/abcf9e56218897d0cf364d4a522da918c689c5aa/src/sno_py/snocommand.py#L69

https://github.com/nitanmarcel/Sno-py/blob/abcf9e56218897d0cf364d4a522da918c689c5aa/src/sno_py/snoedit.py#L328
